### ISSUE TITLE
VRAM leak in Gemma-4 hybrid attention path under activation checkpointing

### ISSUE THREAD

**Issue by thad0ctor** [axolotl-ai-cloud/axolotl#3610]
### Please check that this issue hasn't been reported before.

- [x] I searched previous [Bug Reports](https://github.com/axolotl-ai-cloud/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

VRAM usage should remain relatively flat across training steps when using gemma4_hybrid_attn_impl: true (FA2 for sliding-window layers + SDPA for global layers), matching the flat VRAM profile of the all-flex-attention path but at the ~22× higher throughput the  hybrid path delivers.

### Current behaviour

With gemma4_hybrid_attn_impl: true, VRAM climbs ~0.47 GiB/step from a 42 GiB baseline and OOMs around step 73 (~94 GiB peak).   

The climb is independent of actual sequence length and image size — it happens regardless of input shape variability.
                                                                                                                            
With flex_attention: true (all 60 layers flex), VRAM is flat at 64 GiB but training runs at ~30 s/step vs ~1.3 s/step on hybrid — not a viable workaround.
                                                                                                                                  
Introduced by PR #3598 (commit b8358aa5, "[gemma4] use mixed Flash Attention and SDPA and add fused RMSNorm+RoPE Triton kernels"), which added the hybrid flag, the unconditional patch_gemma4_fused_attn() monkeypatch, and the fused RMSNorm+RoPE kernels together. 

### Steps to reproduce

  1. Install axolotl at HEAD 323da791 (post PR #3598).                                                                            
  2. Launch training on Gemma-4 31B multimodal with the following config elements:
    - gemma4_hybrid_attn_impl: true                                                                                               
    - adapter: lora with lora_r: 256 on the language layers; vision tower + embed_vision trained full                             
    - sample_packing: false                                                                                                       
    - pad_to_sequence_len: false                                                                                                  
    - sequence_len: 5248                                                                                                          
    - FSDP2 with reshard_after_forward: true                      
    - gradient_checkpointing: true                                                                                                
  3. Hardware: 2× RTX PRO 6000 Blackwell 96 GB (also reproducible on any FSDP2 + activation-checkpointing setup per the analysis).
  4. Monitor nvidia-smi or torch.cuda.memory_allocated() per step — observe linear ~0.47 GiB/step climb from the ~42 GiB baseline until OOM around step 73.    

### Config yaml

```yaml
base_model: /home/trainer/model_shims/gemma-4-31B-it-max-1120
tokenizer_type: AutoTokenizer
processor_type: AutoProcessor

seed: 42
data_seed: 42

load_in_8bit: false
load_in_4bit: false
strict: false

adapter: lora
lora_r: 256
lora_alpha: 256          # rsLoRA -> alpha == r is optimal
lora_dropout: 0.05
peft_use_dora: false     # disabled for fast iteration; flip back on for final runs
peft_use_rslora: true    # rank-stabilized LoRA - critical at high rank
peft_autocast_adapter_dtype: false

lora_target_modules: 'model\.language_model\.layers\.[\d]+\.(_checkpoint_wrapped_module\.)?(mlp|self_attn)\.(up|down|gate|q|k|v|o)_proj'

# Optimized LoRA kernels (Liger / KernelsPlugin LoRA_QK for attention_k_eq_v)
#lora_mlp_kernel: true
#lora_qkv_kernel: true
#lora_o_kernel: true

# lora_modules_to_save:
#   - embed_tokens


multimodal: true
skip_prepare_dataset: true
remove_unused_columns: false
sample_packing: false
freeze_mm_modules: false

# Gemma-4 ships its own chat template in axolotl
chat_template: gemma4

default_system_message:
train_on_inputs: false
group_by_length: false

datasets:
  - path: /home/out/train.normalized.jsonl
    type: chat_template

# Held-out validation set, overwritten in place by RUN_FILTER_DATASET.sh
test_datasets:
  - path: /home/out/val.normalized.jsonl
    type: chat_template
    split: train          # local jsonl - axolotl expects this split name

sequence_len: 5248
pad_to_sequence_len: false
eval_sample_packing: false

output_dir: /home/output
num_epochs: 3
micro_batch_size: 1
gradient_accumulation_steps: 4
learning_rate: 0.0001         # default LR for the LoRA A/B adapters on LM layers

embedding_lr: 0.00003         # 3e-5 — conservative, prevents subtle gen quality drift

lr_groups:
  - name: vision_tower
    modules:
      - vision_tower          # catches model.vision_tower.*
    lr: 0.00003               # 3e-5 — adapts for OOD Rashi script without nuking general visual priors
  - name: projector
    modules:
      - embed_vision          # catches model.embed_vision.embedding_projection.*
    lr: 0.00007               # 7e-5 — bridge layer benefits from faster adaptation
lr_scheduler: cosine
warmup_ratio: 0.03
weight_decay: 0.0
label_smoothing_factor: 0.0

max_grad_norm: 1.0
# 8-bit AdamW (torchao) — required to fit full vision FT + r=512 DoRA + embed
# trainable. FSDP2-compatible, ~negligible quality impact for LoRA.
optimizer: adamw_torch_8bit
adam_beta1: 0.9
adam_beta2: 0.999
adam_epsilon: 0.00000001

save_strategy: steps
save_steps: 200
eval_strategy: steps
eval_steps: 200
logging_steps: 10
# gc_steps: 1   # deprecated — replaced by the two fields below (PR #3604)
torch_empty_cache_steps: 10   # HF-native torch.cuda.empty_cache() every 10 steps
gc_collect_steps: 10          # Python gc.collect() every 10 steps
save_total_limit: 12
save_safetensors: true

load_best_model_at_end: true
metric_for_best_model: eval_loss
greater_is_better: false

bf16: true
tf32: true

fp8: false
fp8_enable_fsdp_float8_all_gather: false

gradient_checkpointing: false

dataloader_num_workers: 8
dataloader_pin_memory: true
dataloader_persistent_workers: true
dataloader_prefetch_factor: 8

# =============================================================================
# Attention - use Flex Attention instead of Gemma4 hybrid Flash/SDPA.
# This avoids Flash Attention limits for Gemma4 global_head_dim=512.
# =============================================================================
gemma4_hybrid_attn_impl: true
#flex_attention: true
#flash_attention: true
attention_dropout: 0.0
xformers_attention: false
sdp_attention: false
use_reentrant: false

# =============================================================================
# Plugins - CCE + KernelsPlugin + Liger
# =============================================================================
plugins:
  - axolotl.integrations.cut_cross_entropy.CutCrossEntropyPlugin
  - axolotl.integrations.kernels.KernelsPlugin
  - axolotl.integrations.liger.LigerPlugin

#use_kernels: true       # enables LoRA_QK fused kernel for attention_k_eq_v=true

# Liger fused ops
liger_layer_norm: true
liger_rope: true
liger_rms_norm: true
liger_glu_activation: true
liger_rms_norm_gated: true
liger_fused_linear_cross_entropy: false   # CCE handles the LM-head loss instead

torch_compile: true
compile: true
use_fast: true

# Early stopping
early_stopping_patience: 25

# Gemma-4 is in-tree in transformers >=5.5 — trust_remote_code must be OFF
# because the LoRA fused kernels (lora_mlp_kernel / lora_qkv_kernel /
# lora_o_kernel) are incompatible with custom-code paths.
trust_remote_code: false

# Validation
val_set_size: 0           # disabled - using test_datasets (real held-out val) instead
debug: false
debug_data_preprocessing: false
eval_batch_size: 1

# =============================================================================
# FSDP2 Configuration (2x RTX PRO 6000 Blackwell, 96GB each)
# =============================================================================
fsdp_version: 2
fsdp_config:
  offload_params: false           # 192GB total; keep optimizer on GPU
  # NOTE: cpu_ram_efficient_loading must be false when using FP8 (torchao) +
  # FSDP2 — accelerate explicitly rejects that combo. Each rank will load the
  # full bf16 model into CPU RAM (~62 GB); with 377 GB system RAM that's fine.
  cpu_ram_efficient_loading: false
  auto_wrap_policy: TRANSFORMER_BASED_WRAP
  # Gemma-4 decoder + vision encoder layer classes
  transformer_layer_cls_to_wrap: Gemma4TextDecoderLayer,Gemma4VisionEncoderLayer
  state_dict_type: SHARDED_STATE_DICT
  reshard_after_forward: true
  activation_checkpointing: true
  sync_module_states: true

local_rank:

# Dataset processing
dataset_prepared_path:
push_dataset_to_hub:
hf_use_auth_token:
preprocess_in_chat_template: true
shuffle_merged_datasets: true
dataset_num_proc: 8       # renamed from dataset_processes (deprecated)

# Loss
ignore_index: -100
roles_to_train:
  - assistant
train_on_eos: turn

# Grad accumulation
gradient_accumulation_kwargs:
  sync_each_batch: false

save_model_complete: true
save_best_only: false
```

### Possible solution

**Root cause.** 

HF's Gemma4TextModel.forward allocates one shared_kv_states = {} dict per forward and passes it as a kwarg to every  decoder_layer(...) call. Producer layers (store_full_length_kv=True) write K/V tensors into this dict; consumer layers (is_kv_shared_layer=True) read from it.                                                                                         
                                                                  
  GradientCheckpointingLayer.__call__ forms partial(super().__call__, **kwargs) when checkpointing is active. That partial — which holds a reference to the shared dict — is captured by the checkpoint implementation (axolotl's CPU_Offloaded_Gradient_Checkpointer via ctx.forward_function = forward_function at                        
  src/axolotl/monkeypatch/gradient_checkpointing/offload_cpu.py:51, or torch's stock torch.utils.checkpoint.checkpoint). Because  the dict is a mutable Python container (not an autograd-managed tensor), the K/V tensors it holds are pinned for the full  duration of backward. Combined with delayed ref-cycle cleanup in torch's caching allocator under FSDP2 + activation checkpointing, the residual bleeds across training steps.

  The all-flex path does not leak because it compiles all 60 layers identically into a single graph context where saved tensors are managed by the compiled graph rather than Python refcounting.
                                                                                                                                  
  Violated invariant: anything crossing an activation-checkpoint boundary must be a tensor (refcounted by autograd) or plain Python data — never a mutable container holding tensor references.

  **Possible Solution**                                                                                                               
                                                                  
  Route shared_kv_states through a thread-local side channel instead of the decoder-layer kwargs, so the checkpoint partial never 
  references the dict. All changes in src/axolotl/monkeypatch/models/gemma4/fused_attn.py:
                                                                                                                                  
  1. Add a threading.local() store with _get_shared_kv_states() / _set_shared_kv_states() helpers.
  2. Add _patch_decoder_layer_call() — monkeypatches Gemma4TextDecoderLayer.__call__ to pop shared_kv_states from kwargs and stash it in TLS before delegating to radientCheckpointingLayer.__call__. The partial formed downstream no longer references the dict.
  3. Modify fused_forward to prefer the TLS store, falling back to the kwarg for callers that bypass the patched __call__ (e.g.,  direct attention invocation during eval).                                                                                       
  4. Wire _patch_decoder_layer_call() into patch_gemma4_fused_attn(). Idempotent via a _axolotl_shared_kv_tls_patched sentinel on the class.                                                                                                                      
                                                                  
  TLS is overwritten on each new step's first decoder-layer call, so the previous step's dict is released promptly. No changes to  the hybrid dispatch, FSDP wrap policy, or any config behavior. 
                                                                                                                                  
  **Hypotheses considered and rejected during diagnosis:**                                                                            
  - H1 — per-layer shallow-copied config breaks FSDP/dynamo equivalence: contributes to path divergence but is not the leak  source.                                                                                                                         
  - H2 — silent cfg.flash_attention = True activates downstream per-step allocations: checked all cfg.flash_attention consumers;
  none allocate per-step state on the reported config (sample_packing: false, SM120 skips FA4 auto-patch).                        
  - H4 — FA2 workspace retention under mixed dispatch: FA2 workspace is stateless per-call; no accumulation.                      
   
  **Validation performed:**                                                                                                           
  - ast.parse clean.                                              
  - Patch installs: both Gemma4TextDecoderLayer.__call__ and Gemma4TextAttention.forward are replaced.                            
  - TLS get/set round-trips correctly.                                                                
  - Double-invocation of patch_gemma4_fused_attn() does not double-wrap.                                                          
  - End-to-end VRAM-flatness across steps to be confirmed by reporter.  

  | step | was (pre-patch) | is (post-patch) |                                                                 
  |-----:|---------------:|---------------:|                                                                   
  | eval | 42.5 GiB | 42.6 GiB |                                                                               
  | 9    | 57.2 GiB | 55.3 GiB |                                                                               
  | 19   | 64.8 GiB | 56.3 GiB |                                                                               
  | 29   | 62.0 GiB | 48.7 GiB |                                                                               
  | 39   | 72.5 GiB | 55.0 GiB |
  | 49   | 74.7 GiB | 52.1 GiB |                                                                               
  | 59   | 79.2 GiB | 49.0 GiB |                                  
  | 69   | 85.6 GiB | 52.9 GiB |                                                                               
  | 73   | **OOM (~94 GiB)** | 57.1 GiB (step 79) |                                                            
  | 89   | — | 52.7 GiB |                                                                                      
  | 99   | — | 57.5 GiB |                                                                                      
  | 109  | — | 48.5 GiB |                                                                                      
  | 119  | — | 53.4 GiB |                                                                                      
  
  **Pre-patch:** monotonic climb ~0.47 GiB/step → OOM at step 73.                                              
  **Post-patch:** oscillates in a 48–58 GiB band with no trend across 120+ steps. Peak at step 99 (57.5 GiB)
  matches early-run peak at step 19 (56.3 GiB). 

### Which Operating Systems are you using?

- [x] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.12

### axolotl branch-commit

main/323da791

### Acknowledgements

- [x] My issue title is concise, descriptive, and in title casing.
- [x] I have searched the existing issues to make sure this bug has not been reported yet.
- [x] I am using the latest version of axolotl.
- [x] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by thad0ctor** [comment]
Addressed via https://github.com/axolotl-ai-cloud/axolotl/pull/3611, currently >300 steps of training completed (vs ~70 before OOM) with stable VRAM across training
